include/cuco/utility/fast_int.cuh

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuco/detail/__config>
+
+#include <cstdint>
+#include <cuda/std/bit>
+#include <cuda/std/limits>
+#include <cuda/std/type_traits>
+
+namespace cuco::utility {
+
+/**
+ * @brief Integer type with optimized division and modulo operators.
+ *
+ * @tparam T Underlying integer type
+ */
+template <typename T>
+struct fast_int {
+  static_assert(cuda::std::is_same_v<T, std::int32_t> or cuda::std::is_same_v<T, std::uint32_t>
+#if defined(CUCO_HAS_INT128)
+                  or cuda::std::is_same_v<T, std::int64_t> or cuda::std::is_same_v<T, std::uint64_t>
+#endif
+                ,
+                "Unsupported type");
+
+  using value_type = T;  ///< Underlying integer type
+
+  /**
+   * @brief Constructs a fast_int from an integer value.
+   *
+   * @param value Integer value
+   */
+  __host__ __device__ constexpr fast_int(T value) noexcept : value_{value}
+  {
+    evaluate_magic_numbers();
+  }
+
+  /**
+   * @brief Implicit conversion operator to the underlying value type.
+   *
+   * @return Underlying value
+   */
+  __host__ __device__ constexpr operator value_type() const noexcept { return value_; }
```

Comment:
```suggestion
  __host__ __device__ explicit constexpr operator value_type() const noexcept { return value_; }
```

Reply:
Implicit conversions like this almost always cause pain later down the road.

Reply:
Can you give an example where this could go wrong? The worst I can come up with is that the compiler decides to call the builtin `operator/` and `operator%` over the optimized implementations. I've double-checked that this isn't the case.

Reply:
Nevermind, I already ran into an issue with `int32_t(42) % fast_int<int64_t>(7)` resulting in an operator ambiguity:
```
warning: ISO C++ says that these are ambiguous, even though the worst conversion for the first is better than the worst conversion for the second
```

Jason Turner was right. Implicit conversion are pure evil.

Reply:
The pain isn't just limited to `operator/` or `operator%`. You open yourself up to a whole world of pain.

```
void foo(int);
...

fast_int i;
foo(i); // whoops, did I mean to do that?
```

Reply:
Fixed in [5076c33](https://github.com/NVIDIA/cuCollections/pull/309/commits/5076c3300ce87791046327f8edc37936ee17b21b)